<a href="https://colab.research.google.com/github/emmanuelwanyama/INTERNSHIP_2023/blob/main/Workshop_2/Copy_of_Workshop_2_Module_1_Introduction_to_pandas_DataFrames_2022_03_16.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![banner](https://github.com/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io/blob/master/docs/banner.jpg?raw=1)

***[Workshop 2](about) - Training course in data analysis for genomic surveillance of African malaria vectors***

---

# Module 1 - Introduction to pandas DataFrames

**Theme: Tools & Technology**

In this module we're going to take a closer look at a Python package called [pandas](https://pandas.pydata.org/), which is an essential part of our data analysis toolkit. We use pandas whenever we need to store, manipulate or analyse tables of data, also known as DataFrames. In this module we'll focus on using pandas to analyse genomic data, but it can be used to analyse any kind of data, so it's a very useful tool to have at your fingertips.      

## Learning objectives

After this module you will be able to:

* Explain what a DataFrame is
* Access columns and rows from a DataFrame
* Use indexes and queries
* Write a DataFrame to a file

Please note that this module will cover only a small part of the features that pandas provides - see the links at then end of this notebook if you'd like to go further.

## Lecture

### English

In [ ]:
%%html
<iframe width="560" height="315" src="https://www.youtube.com/embed/iIYftkvS0eM" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

### Français

In [ ]:
%%html
<iframe width="560" height="315" src="https://www.youtube.com/embed/RdyXA31iHz8" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>


Please note that the code in the cells below might differ from that shown in the video. This can happen because Python packages and their dependencies change due to updates, necessitating tweaks to the code.

## Setup

Let's install the `malariagen_data` package so we have easy access to some genomic data to demonstrate with. Pandas is already installed on colab.

In [ ]:
!pip install -q --no-warn-conflicts malariagen_data

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 131.7/131.7 kB 2.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 13.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 35.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.4/10.4 MB 65.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 33.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.5/302.5 kB 26.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 86.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 138.7/138.7 kB 15.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.9/20.9 MB 24.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.4/7.4 MB 33.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 206.5/206.5 kB 21.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.7/56.7 kB 6.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

Now import the `malariagen_data` and `pandas` packages. Note that by convention we use `pd` as an abbreviation for `pandas`.

In [ ]:
import malariagen_data
import pandas as pd

Set up access to the MalariaGEN Ag3.0 data resource.

In [ ]:
ag3 = malariagen_data.Ag3()
ag3

<MalariaGEN Ag3 API client>
Storage URL             : gs://vo_agam_release/
Data releases available : 3.0
Results cache           : None
Cohorts analysis        : 20230516
AIM analysis            : 20220528
Site filters analysis   : dt_20200416
Software version        : malariagen_data 7.12.0
Client location         : Nevada, US
---
Please note that data are subject to terms of use,
for more information see https://www.malariagen.net/data
or contact data@malariagen.net. For API documentation see 
https://malariagen.github.io/vector-data/ag3/api.html

## pandas DataFrames

We use the pandas `DataFrame` class whenever we want to store and analyse data that can be structured as a table with rows and columns. Let's take a look at some examples of DataFrames that we met in the [first workshop](https://anopheles-genomic-surveillance.github.io/workshop-1/about.html) of this training course.

### Example: sample metadata

In [workshop 1, module 2](https://anopheles-genomic-surveillance.github.io/workshop-1/module-2-sample-metadata.html) we encountered the [sample_metadata()](https://malariagen.github.io/vector-data/ag3/api.html#sample-metadata) function. This function returns a DataFrame of sample metadata. Let's call this function and assign the result to the variable `df_samples`.

In [ ]:
df_samples = ag3.sample_metadata(sample_sets="3.0")

Check what type of object we have received.

In [ ]:
type(df_samples)

pandas.core.frame.DataFrame

Inspect the data in the DataFrame by running a code cell containing just the name of the variable.

In [ ]:
df_samples

sample_id partner_sample_id      contributor    country location  year  \
0     AR0047-C            LUA047       Joao Pinto     Angola   Luanda  2009   
1     AR0049-C            LUA049       Joao Pinto     Angola   Luanda  2009   
2     AR0051-C            LUA051       Joao Pinto     Angola   Luanda  2009   
3     AR0061-C            LUA061       Joao Pinto     Angola   Luanda  2009   
4     AR0078-C            LUA078       Joao Pinto     Angola   Luanda  2009   
...        ...               ...              ...        ...      ...   ...   
3076  AD0494-C         80-2-o-16  Martin Donnelly  Lab Cross     LSTM    -1   
3077  AD0495-C         80-2-o-17  Martin Donnelly  Lab Cross     LSTM    -1   
3078  AD0496-C         80-2-o-18  Martin Donnelly  Lab Cross     LSTM    -1   
3079  AD0497-C         80-2-o-19  Martin Donnelly  Lab Cross     LSTM    -1   
3080  AD0498-C         80-2-o-20  Martin Donnelly  Lab Cross     LSTM    -1   

      month  latitude  longitude sex_call  ... admin1_name admin1_iso  \
0         4    -8.884     13.302        F  ...      Luanda     AO-LUA   
1         4    -8.884     13.302        F  ...      Luanda     AO-LUA   
2         4    -8.884     13.302        F  ...      Luanda     AO-LUA   
3         4    -8.884     13.302        F  ...      Luanda     AO-LUA   
4         4    -8.884     13.302        F  ...      Luanda     AO-LUA   
...     ...       ...        ...      ...  ...         ...        ...   
3076     -1    53.409     -2.969        F  ...         NaN        NaN   
3077     -1    53.409     -2.969        M  ...         NaN        NaN   
3078     -1    53.409     -2.969        M  ...         NaN        NaN   
3079     -1    53.409     -2.969        F  ...         NaN        NaN   
3080     -1    53.409     -2.969        M  ...         NaN        NaN   

      admin2_name       taxon  cohort_admin1_year  cohort_admin1_month  \
0          Luanda    coluzzii    AO-LUA_colu_2009  AO-LUA_colu_2009_04   
1          Luanda    coluzzii    AO-LUA_colu_2009  AO-LUA_colu_2009_04   
2          Luanda    coluzzii    AO-LUA_colu_2009  AO-LUA_colu_2009_04   
3          Luanda    coluzzii    AO-LUA_colu_2009  AO-LUA_colu_2009_04   
4          Luanda    coluzzii    AO-LUA_colu_2009  AO-LUA_colu_2009_04   
...           ...         ...                 ...                  ...   
3076          NaN  unassigned                 NaN                  NaN   
3077          NaN  unassigned                 NaN                  NaN   
3078          NaN  unassigned                 NaN                  NaN   
3079          NaN  unassigned                 NaN                  NaN   
3080          NaN  unassigned                 NaN                  NaN   

     cohort_admin1_quarter       cohort_admin2_year  \
0      AO-LUA_colu_2009_Q2  AO-LUA_Luanda_colu_2009   
1      AO-LUA_colu_2009_Q2  AO-LUA_Luanda_colu_2009   
2      AO-LUA_colu_2009_Q2  AO-LUA_Luanda_colu_2009   
3      AO-LUA_colu_2009_Q2  AO-LUA_Luanda_colu_2009   
4      AO-LUA_colu_2009_Q2  AO-LUA_Luanda_colu_2009   
...                    ...                      ...   
3076                   NaN                      NaN   
3077                   NaN                      NaN   
3078                   NaN                      NaN   
3079                   NaN                      NaN   
3080                   NaN                      NaN   

             cohort_admin2_month       cohort_admin2_quarter  
0     AO-LUA_Luanda_colu_2009_04  AO-LUA_Luanda_colu_2009_Q2  
1     AO-LUA_Luanda_colu_2009_04  AO-LUA_Luanda_colu_2009_Q2  
2     AO-LUA_Luanda_colu_2009_04  AO-LUA_Luanda_colu_2009_Q2  
3     AO-LUA_Luanda_colu_2009_04  AO-LUA_Luanda_colu_2009_Q2  
4     AO-LUA_Luanda_colu_2009_04  AO-LUA_Luanda_colu_2009_Q2  
...                          ...                         ...  
3076                         NaN                         NaN  
3077                         NaN                         NaN  
3078                         NaN                         NaN  

There are 3,081 rows and 26 columns in this DataFrame. Each row provides data about a single mosquito DNA sample that has been sequenced by MalariaGEN. The columns provide data about those mosquitoes, such as the country and year in which the mosquito was originally collected.

Note that the representation of the DataFrame above only shows some of the data - there are too many rows and columns to show all at once. The ellipsis character ("...") shows where some data have been hidden.

### Example: genome annotations

In [workshop 1, module 3](https://anopheles-genomic-surveillance.github.io/workshop-1/module-3-genome.html) we encountered the [geneset()](https://malariagen.github.io/vector-data/ag3/api.html#geneset) function. This function returns a DataFrame of genome annotations. Let's call the function and assign the result to the variable `df_geneset`.

In [ ]:
df_geneset = ag3.geneset()
df_geneset

contig      source             type   start       end  score  \
0               2L  VectorBase       chromosome       1  49364325    NaN   
1               2L  VectorBase             gene  157348    186936    NaN   
2               2L  VectorBase             mRNA  157348    181305    NaN   
3               2L  VectorBase  three_prime_UTR  157348    157495    NaN   
4               2L  VectorBase             exon  157348    157623    NaN   
...            ...         ...              ...     ...       ...    ...   
196140  Y_unplaced  VectorBase   five_prime_UTR   47932     48111    NaN   
196141  Y_unplaced  VectorBase             exon   47932     48138    NaN   
196142  Y_unplaced  VectorBase              CDS   48112     48138    NaN   
196143  Y_unplaced  VectorBase             exon   48301     48385    NaN   
196144  Y_unplaced  VectorBase              CDS   48301     48385    NaN   

       strand  phase             ID         Parent              Name  \
0         NaN    NaN             2L            NaN               NaN   
1           -    NaN     AGAP004677            NaN               NaN   
2           -    NaN  AGAP004677-RA     AGAP004677               NaN   
3           -    NaN            NaN  AGAP004677-RA               NaN   
4           -    NaN            NaN  AGAP004677-RA  AGAP004677-RB-E4   
...       ...    ...            ...            ...               ...   
196140      +    NaN            NaN  AGAP029375-RA               NaN   
196141      +    NaN            NaN  AGAP029375-RA  AGAP029375-RA-E2   
196142      +    0.0  AGAP029375-PA  AGAP029375-RA               NaN   
196143      +    NaN            NaN  AGAP029375-RA  AGAP029375-RA-E3   
196144      +    0.0  AGAP029375-PA  AGAP029375-RA               NaN   

                                              description  
0                                                     NaN  
1       methylenetetrahydrofolate dehydrogenase(NAD ) ...  
2                                                     NaN  
3                                                     NaN  
4                                                     NaN  
...                                                   ...  
196140                                                NaN  
196141                                                NaN  
196142                                                NaN  
196143                                                NaN  
196144                                                NaN  

[196145 rows x 12 columns]

There are 196,145 rows and 12 columns in this DataFrame. Each row provides data about an annotated feature on the *Anopheles gambiae* AgamP4 reference genome, such as a gene. The columns provide data about those features, such as the contig (chromosome arm) on which the feature occurs, and the start and end coordinates.

### Example: SNP allele frequencies

In [workshop 1, module 4](https://anopheles-genomic-surveillance.github.io/workshop-1/module-4-vgsc-snps.html), we encountered the [snp_allele_frequencies()](https://malariagen.github.io/vector-data/ag3/api.html#snp-allele-frequencies) function. This function returns a DataFrame of single nucleotide polymorphisms (SNPs) together with their allele frequencies in different mosquito cohorts. Let's call this function and assign the result to the variable `snp_freqs_df`.

In [ ]:
snp_freqs_df = ag3.snp_allele_frequencies(
    transcript="AGAP004707-RD",
    cohorts="admin1_year",
    sample_sets=["AG1000G-BF-A", "AG1000G-BF-B", "AG1000G-BF-C"],
)
snp_freqs_df

Load SNP genotypes:   0%|          | 0/32 [00:00<?, ?it/s]

Compute allele frequencies:   0%|          | 0/6 [00:00<?, ?it/s]

Compute SNP effects:   0%|          | 0/9675 [00:00<?, ?it/s]

pass_gamb_colu_arab  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                      True   
       2358328  T          C          NaN                       True   
       2358441  A          T          NaN                       True   
       2358468  A          C          NaN                       True   
       2358581  T          A          NaN                       True   
...                                                              ...   
       2431424  G          T          G2055V                    True   
       2431518  C          A          A2086A                    True   
       2431527  G          C          V2089V                    True   
       2431569  C          A          V2103V                    True   
       2431613  T          C          V2118A                    True   

                                                 pass_gamb_colu  pass_arab  \
contig position ref_allele alt_allele aa_change                              
2L     2358252  C          T          A32V                 True       True   
       2358328  T          C          NaN                  True       True   
       2358441  A          T          NaN                  True       True   
       2358468  A          C          NaN                  True       True   
       2358581  T          A          NaN                  True       True   
...                                                         ...        ...   
       2431424  G          T          G2055V               True       True   
       2431518  C          A          A2086A               True       True   
       2431527  G          C          V2089V               True       True   
       2431569  C          A          V2103V               True       True   
       2431613  T          C          V2118A               True       True   

                                                 frq_BF-07_gamb_2004  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                  0.000000   
       2358328  T          C          NaN                   0.000000   
       2358441  A          T          NaN                   0.038462   
       2358468  A          C          NaN                   0.269231   
       2358581  T          A          NaN                   0.192308   
...                                                              ...   
       2431424  G          T          G2055V                0.000000   
       2431518  C          A          A2086A                0.038462   
       2431527  G          C          V2089V                0.038462   
       2431569  C          A          V2103V                0.038462   
       2431613  T          C          V2118A                0.000000   

                                                 frq_BF-09_colu_2012  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                  0.006098   
       2358328  T          C          NaN                   0.006098   
       2358441  A          T          NaN                   0.000000   
       2358468  A          C          NaN                   0.000000   
       2358581  T          A          NaN                   0.000000   
...                                                              ...   
       2431424  G          T          G2055V                0.000000   
       2431518  C          A          A2086A                0.000000   
       2431527  G          C          V2089V                0.000000   
       2431569  C          A          V2103V                0.000000   
       2431613  T          C          V2118A                0.000000   

                                                 frq_BF-09_colu_2014  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                       0.0   
       

There are 9,675 rows and 18 columns in this DataFrame. Each row provides data about a SNP. The columns provide data about those SNPs, such as their effect, and the alternate allele frequency in a particular cohort of mosquitoes, e.g., the column named "frq_BF-09_colu_2012" provides the frequency in *Anopheles coluzzii* mosquitoes collected in the BF-09 (Hauts-Bassins) administrative region of Burkina Faso in 2012.

### Aside: variable naming styles

You may have noticed that in the examples above, whenever we assigned a DataFrame to a variable, the name of the variable either started with the prefix "`df_`" or ended with the suffix "`_df`". Here "df" stands for DataFrame, and is used just as a helpful reminder that the value of the variable is a pandas DataFrame. Using a prefix of suffix on the variable name like this is entirely optional and is up to your own coding style. You can name these variables whatever you like. Different instructors in this course prefer different styles, and you will see different styles being used in different modules.

### Example: reading from a CSV file

You can also create a pandas DataFrame by reading data from a file, such as a comma-separated value (CSV) file, or an Excel workbook file. Let's use the pandas [read_csv()](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) function to read data from a CSV file containing data from a US census that colab provides as example data for you to play with.

In [ ]:
df_housing = pd.read_csv("sample_data/california_housing_train.csv")
df_housing

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -114.31     34.19                15.0       5612.0          1283.0   
1        -114.47     34.40                19.0       7650.0          1901.0   
2        -114.56     33.69                17.0        720.0           174.0   
3        -114.57     33.64                14.0       1501.0           337.0   
4        -114.57     33.57                20.0       1454.0           326.0   
...          ...       ...                 ...          ...             ...   
16995    -124.26     40.58                52.0       2217.0           394.0   
16996    -124.27     40.69                36.0       2349.0           528.0   
16997    -124.30     41.84                17.0       2677.0           531.0   
16998    -124.30     41.80                19.0       2672.0           552.0   
16999    -124.35     40.54                52.0       1820.0           300.0   

       population  households  median_income  median_house_value  
0          1015.0       472.0         1.4936             66900.0  
1          1129.0       463.0         1.8200             80100.0  
2           333.0       117.0         1.6509             85700.0  
3           515.0       226.0         3.1917             73400.0  
4           624.0       262.0         1.9250             65500.0  
...           ...         ...            ...                 ...  
16995       907.0       369.0         2.3571            111400.0  
16996      1194.0       465.0         2.5179             79000.0  
16997      1244.0       456.0         3.0313            103600.0  
16998      1298.0       478.0         1.9797             85800.0  
16999       806.0       270.0         3.0147             94600.0  

[17000 rows x 9 columns]

This DataFrame has 17,000 rows and 9 columns. Each row provides data about a block of housing. The columns provide data about the houses within a block, such as the total number of bedrooms and the number of households.

Note that you can upload your own files to Colab if you want to read in and analyse some of your own data with pandas.

### Example: creating a DataFrame from scratch

You can also create a DataFrame by providing the data within your code. To demonstrate this, let's create a DataFrame with data about characters in a famous science fiction book called The Hitchhiker's Guide to the Galaxy.

In [ ]:
df_hitchhikers = pd.DataFrame(data={
    "name": ["Arthur Dent", "Tricia McMillan", "Ford Prefect", "Zaphod Beeblebrox", "Marvin the Paranoid Android"],
    "taxon": ["human", "human", "betelgeusian", "betelgeusian", "robot"],
    "gender": ["male", "female", "male", "male", None],
    "number_of_heads": [1, 1, 1, 2, 1],
    "intelligence": [100, 180, 120, 30, 5_000_000]
})
df_hitchhikers

name         taxon  gender  number_of_heads  \
0                  Arthur Dent         human    male                1   
1              Tricia McMillan         human  female                1   
2                 Ford Prefect  betelgeusian    male                1   
3            Zaphod Beeblebrox  betelgeusian    male                2   
4  Marvin the Paranoid Android         robot    None                1   

   intelligence  
0           100  
1           180  
2           120  
3            30  
4       5000000

This DataFrame has 5 rows and 5 columns. Each row provides data about a fictional character in the book. The columns provide data about each character, such as their name, taxon and number of heads (some alien species can have more than one).

## Structure of a DataFrame

Hopefully this will already be obvious from the examples above, but a [pandas DataFrame](https://pandas.pydata.org/docs/user_guide/dsintro.html#dataframe) is a data structure that is organised into **rows** and **columns**. Each DataFrame also has an **index**, which is a special column (or set of columns) that is shown in bold at the left of a DataFrame, and which we'll look at in more detail later.

In [ ]:
%%html
<img width="90%" height="90%"src="https://storage.googleapis.com/vo_agam_release/reference/training-images/workshop-2/W2M1-1.PNG"/>

## Counting rows and columns

If a DataFrame is large, you may find it useful to know how many rows and columns it contains. The representation of a DataFrame shows at the bottom how many rows and columns are present in the DataFrame, but you can also use the built-in `len()` function if you want to access the number of rows or columns in your code.

In [ ]:
num_rows = len(df_hitchhikers)
num_rows

5

In [ ]:
num_cols = len(df_hitchhikers.columns)
num_cols

5

**Exercise 1 (English)**

* Count the number of rows and columns in the `df_samples` DataFrame.

**Exercice 1 (Français)**

* Compter le nombre de lignes et de colonnes du DataFrame `df_samples`.

In [ ]:
num_rows, num_cols = df_samples.shape
print(f"Number of rows in df_samples: {num_rows}")
print(f"Number of columns in df_samples: {num_cols}")


Number of rows in df_samples: 3081
Number of columns in df_samples: 30


## Accessing columns

Often you will have a DataFrame with many columns, and you many only need to work with data from a smaller number of columns, or just a single column. Accessing one or more columns from a DataFrame can be done using the square bracket syntax after the name of a DataFrame variable. Let's see how this works.


### Accessing column names

First, let's access the names of all the columns in a DataFrame, to see what columns are available. You can do this by accessing the `columns` property of a DataFrame. For simplicity, let's also convert the result to a Python list, using the built-in `list()` function.

In [ ]:
col_names = list(df_hitchhikers.columns)
col_names

['name', 'taxon', 'gender', 'number_of_heads', 'intelligence']

This can be useful especially if you have a lot of columns in a DataFrame.

**Exercise 2 (English)**

* Access the column names of the `df_samples` DataFrame.

**Exercice 2 (Français)**

* Afficher les noms des colonnes du DataFrame `df_samples`.

In [ ]:
column_names = df_samples.columns
print(column_names)


Index(['sample_id', 'partner_sample_id', 'contributor', 'country', 'location',
       'year', 'month', 'latitude', 'longitude', 'sex_call', 'sample_set',
       'release', 'quarter', 'aim_species_fraction_arab',
       'aim_species_fraction_colu', 'aim_species_fraction_colu_no2l',
       'aim_species_gambcolu_arabiensis', 'aim_species_gambiae_coluzzii',
       'aim_species', 'country_iso', 'admin1_name', 'admin1_iso',
       'admin2_name', 'taxon', 'cohort_admin1_year', 'cohort_admin1_month',
       'cohort_admin1_quarter', 'cohort_admin2_year', 'cohort_admin2_month',
       'cohort_admin2_quarter'],
      dtype='object')


## In a list form

In [ ]:
column_names_list = list(df_samples.columns)
print(column_names_list)


['sample_id', 'partner_sample_id', 'contributor', 'country', 'location', 'year', 'month', 'latitude', 'longitude', 'sex_call', 'sample_set', 'release', 'quarter', 'aim_species_fraction_arab', 'aim_species_fraction_colu', 'aim_species_fraction_colu_no2l', 'aim_species_gambcolu_arabiensis', 'aim_species_gambiae_coluzzii', 'aim_species', 'country_iso', 'admin1_name', 'admin1_iso', 'admin2_name', 'taxon', 'cohort_admin1_year', 'cohort_admin1_month', 'cohort_admin1_quarter', 'cohort_admin2_year', 'cohort_admin2_month', 'cohort_admin2_quarter']


### Accessing a single column

To access a single column, give the column name as a string inside square brackets. Let's access the "name" column from the DataFrame of book characters, and assign the result to a new variable `character_name`.

In [ ]:
character_name = df_hitchhikers["name"]

When you access a single column or row from a DataFrame, pandas returns a [Series](https://pandas.pydata.org/docs/user_guide/dsintro.html#series) object to you. Pandas uses the Series class to hold data either from a single column or a single row, i.e., a one-dimensional sequence of data values.

In [ ]:
type(character_name)

pandas.core.series.Series

We can also inspect the data in the column.

In [ ]:
character_name

0                    Arthur Dent
1                Tricia McMillan
2                   Ford Prefect
3              Zaphod Beeblebrox
4    Marvin the Paranoid Android
Name: name, dtype: object

### Investigating data in a column

The pandas Series class has some useful functions for investigating the data in a DataFrame.

E.g., let's access the "taxon" column and use the `value_counts()` functions to get a summary of the categorical data in the column.

In [ ]:
character_taxon = df_hitchhikers["taxon"]
character_taxon.value_counts()

human           2
betelgeusian    2
robot           1
Name: taxon, dtype: int64

The pandas Series class also has a `describe()` function which can be useful for investigating numerical data. Let's use it with the `number_of_heads` column.

In [ ]:
character_heads = df_hitchhikers["number_of_heads"]
character_heads.describe()

count    5.000000
mean     1.200000
std      0.447214
min      1.000000
25%      1.000000
50%      1.000000
75%      1.000000
max      2.000000
Name: number_of_heads, dtype: float64

The characters in our DataFrame have at least (min) 1 and at most (max) 2 heads. The mean number of heads is 1.2, but of course this a bit silly, because number of heads is not a continuous variable.

**Exercise 3 (English)**

* Access the "country" column from the `df_samples` DataFrame, and then use the `value_counts()` function to show the number of mosquito samples from each country.

* Access the "type" column from the `df_geneset` DataFrame, and then use the `value_counts()` function to show the number of genome annotations of different types.

* Access the "effect" column from the `snp_freqs_df` DataFrame, and then use the `value_counts()` function to show the number of SNPs with different effects.

**Exercice 3 (Français)**

* Obtenir la colonne "country" du DataFrame `df_samples` et utiliser la fonction `value_counts()` pour afficher le nombre de moustiques capturés dans chaque pays.

* Obtenir la colonne "type" du DataFrame `df_geneset` et utiliser la fonction `value_counts()` pour afficher le nombre d'annotations du génome de chaque type.

* Obtenir la colonne "effect" du DataFrame `snp_freqs_df` et utiliser la fonction `value_counts()` pour afficher le nombre de SNPs pour chaque effet possible.

### Access the country column from the df_samples DataFrame, and then use the value_counts() function to show the number of mosquito samples from each country.

In [ ]:
country_counts = df_samples["country"].value_counts()
print(country_counts)


Cameroon                            444
Tanzania                            300
Lab Cross                           297
Burkina Faso                        296
Uganda                              290
Gambia, The                         279
Mali                                225
Guinea                              136
Guinea-Bissau                       101
Ghana                               100
Kenya                                86
Angola                               81
Cote d'Ivoire                        80
Democratic Republic of the Congo     76
Mozambique                           74
Central African Republic             73
Gabon                                69
Malawi                               41
Mayotte                              23
Equatorial Guinea                    10
Name: country, dtype: int64


### Access the "type" column from the df_geneset DataFrame, and then use the value_counts() function to show the number of genome annotations of different types.

In [ ]:
type_counts = df_geneset["type"].value_counts()
print(type_counts)


exon                      72257
CDS                       66103
five_prime_UTR            16598
mRNA                      14979
gene                      13057
three_prime_UTR           11665
ncRNA_gene                  729
tRNA                        362
rRNA                        242
pre_miRNA                    77
snRNA                        35
pseudogene                   10
pseudogenic_transcript       10
chromosome                    8
ncRNA                         4
SRP_RNA                       3
snoRNA                        2
lnc_RNA                       2
RNase_MRP_RNA                 1
RNase_P_RNA                   1
Name: type, dtype: int64


### Access the "effect" column from the snp_freqs_df DataFrame, and then use the value_counts() function to show the number of SNPs with different effects.

In [ ]:
effect_counts = snp_freqs_df["effect"].value_counts()
print(effect_counts)


INTRONIC                 9467
NON_SYNONYMOUS_CODING     121
SYNONYMOUS_CODING          56
STOP_GAINED                14
SPLICE_REGION               9
SPLICE_CORE                 8
Name: effect, dtype: int64


### Accessing multiple columns

To access multiple columns, provide the column names as a list of strings. Let's access the "name", "taxon" and "number_of_heads" columns from the DataFrame of book characters, and assign the resulting DataFrame to a new variable `df_hh`.

In [ ]:
selected_cols = ["name", "taxon", "number_of_heads"]
df_hh = df_hitchhikers[selected_cols]
df_hh

name         taxon  number_of_heads
0                  Arthur Dent         human                1
1              Tricia McMillan         human                1
2                 Ford Prefect  betelgeusian                1
3            Zaphod Beeblebrox  betelgeusian                2
4  Marvin the Paranoid Android         robot                1

You can also put the list of columns directly inside the square brackets, which would give the same result.

In [ ]:
df_hh = df_hitchhikers[["name", "taxon", "number_of_heads"]]
df_hh

name         taxon  number_of_heads
0                  Arthur Dent         human                1
1              Tricia McMillan         human                1
2                 Ford Prefect  betelgeusian                1
3            Zaphod Beeblebrox  betelgeusian                2
4  Marvin the Paranoid Android         robot                1

**Exercise 4 (English)**

* Create a new DataFrame containing only the "sample_id", "country" and "year" columns from the `df_samples` DataFrame.
* Create a new DataFrame containing only the "ID", "contig", "start" and "end" columns from the `df_geneset` DataFrame.
* Create a new DataFrame containing only the frequency columns from the `snp_freqs_df` DataFrame (hint: the names of these columns all begin with "frq_").

**Exercice 4 (Français)**

* Créer un nouveau DataFrame contenant seulement les colonnes "sample_id", "country" et "year" du DataFrame `df_samples`.

* Créer un nouveau DataFrame contenant seulement les colonnes "ID", "contig", "start" et "end" du DataFrame `df_geneset`.

* Créer un nouveau DataFrame contenant seulement les colonnes contenant les fréquences du DataFrame `snp_freqs_df` (Indice: les noms de ces colonnes commencent tous par "frq_").

### Create a new DataFrame containing only the "sample_id", "country" and "year" columns from the df_samples DataFrame.

In [ ]:
# Create a new DataFrame with selected columns
selected_columns_df = df_samples.loc[:, ["sample_id", "country", "year"]]

# Display the new DataFrame
print(selected_columns_df)


     sample_id    country  year
0     AR0047-C     Angola  2009
1     AR0049-C     Angola  2009
2     AR0051-C     Angola  2009
3     AR0061-C     Angola  2009
4     AR0078-C     Angola  2009
...        ...        ...   ...
3076  AD0494-C  Lab Cross    -1
3077  AD0495-C  Lab Cross    -1
3078  AD0496-C  Lab Cross    -1
3079  AD0497-C  Lab Cross    -1
3080  AD0498-C  Lab Cross    -1

[3081 rows x 3 columns]


### Create a new DataFrame containing only the "ID", "contig", "start" and "end" columns from the df_geneset DataFrame.

In [ ]:
# Create a new DataFrame with selected columns
selected_columns_df = df_geneset.loc[:, ["ID", "contig", "start", "end"]]

# Display the new DataFrame
print(selected_columns_df)


                   ID      contig   start       end
0                  2L          2L       1  49364325
1          AGAP004677          2L  157348    186936
2       AGAP004677-RA          2L  157348    181305
3                 NaN          2L  157348    157495
4                 NaN          2L  157348    157623
...               ...         ...     ...       ...
196140            NaN  Y_unplaced   47932     48111
196141            NaN  Y_unplaced   47932     48138
196142  AGAP029375-PA  Y_unplaced   48112     48138
196143            NaN  Y_unplaced   48301     48385
196144  AGAP029375-PA  Y_unplaced   48301     48385

[196145 rows x 4 columns]


### Create a new DataFrame containing only the frequency columns from the snp_freqs_df DataFrame (hint: the names of these columns all begin with "frq_").

In [ ]:
# Filter the DataFrame to select columns that start with "frq_"
frequency_columns_df = snp_freqs_df.filter(regex=r'^frq_')

# Display the new DataFrame
print(frequency_columns_df)


                                                 frq_BF-07_gamb_2004  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                  0.000000   
       2358328  T          C          NaN                   0.000000   
       2358441  A          T          NaN                   0.038462   
       2358468  A          C          NaN                   0.269231   
       2358581  T          A          NaN                   0.192308   
...                                                              ...   
       2431424  G          T          G2055V                0.000000   
       2431518  C          A          A2086A                0.038462   
       2431527  G          C          V2089V                0.038462   
       2431569  C          A          V2103V                0.038462   
       2431613  T          C          V2118A                0.000000   

                                                 frq_BF-09_colu

## Accessing rows

Now we've seen how to access columns, let's see how to access rows. To access rows by their position within a DataFrame, we can use the `iloc` property.



### Accessing a single row

To access a single row, we use the `iloc` property, then provide the row number as an integer inside square brackets, remembering that numbering starts from 0 in Python. For example, access the first row of our DataFrame of book characters, and assign the result to a variable `first_row`.

In [ ]:
first_row = df_hitchhikers.iloc[0]

When accessing a single row, pandas returns a Series.

In [ ]:
type(first_row)

pandas.core.series.Series

Inspect the data in the row.

In [ ]:
first_row

name               Arthur Dent
taxon                    human
gender                    male
number_of_heads              1
intelligence               100
Name: 0, dtype: object

Access specific data values within the row.

In [ ]:
first_row["name"]

'Arthur Dent'

In [ ]:
first_row["number_of_heads"]

1

**Exercise 5 (English)**

* Access the 81st row in the `df_samples` DataFrame. What country did this mosquito sample come from? (Remember: numbering starts from 0 in Python.)

* Access the 73,287th row in the `df_geneset` DataFrame. What type of annotation is this? What is its Name?

**Exercice 5 (Français)**

* Obtenir la 81ème ligne du DataFrame `df_samples`. De quel pays est issu ce moustique? (Indice: les indices commencent à 0 en Python)

* Obtenir la 73 287ème ligne du DataFrame `df_geneset`. De quel type d'annotation s'agit-il? Quel est son nom?

### Access the 81st row in the df_samples DataFrame. What country did this mosquito sample come from? (Remember: numbering starts from 0 in Python.)

In [ ]:
# Access the 81st row in the df_samples DataFrame
row_81 = df_samples.iloc[80]

# Extract the country from the row
country_of_sample = row_81['country']

# Print the country
print(country_of_sample)


Angola


### Access the 73,287th row in the df_geneset DataFrame. What type of annotation is this? What is its Name?

In [ ]:
# Accesd the 73,287th row in the df_geneset DataFrame
row_73287 = df_geneset.iloc[73286]

# Extracted the type of annotation from the row
annotation_type = row_73287['type']

# Extract the Name from the row
annotation_name = row_73287['Name']

# Print the type and Name
print("Type of annotation:", annotation_type)
print("Name of annotation:", annotation_name)


Type of annotation: gene
Name of annotation: CYP6AA1


### Accessing multiple rows

We can also use `iloc` to access multiple rows,from some start position to some end position. To do this we provide two integers separated by a colon (":") which is called a slice.

As well as remembering that numbering starts from zero, we also have to remember that slices include the start position but do not include the stop position. E.g., to access the first two rows, provide the start and stop positions 0 and 2. This returns a new DataFrame with just the selected rows within the slice.

In [ ]:
first_two_rows = df_hitchhikers.iloc[0:2]
first_two_rows

name  taxon  gender  number_of_heads  intelligence
0      Arthur Dent  human    male                1           100
1  Tricia McMillan  human  female                1           180

**Exercise 6 (English)**

* Access rows starting from the 6th row up to and including the 10th row in the `df_samples` DataFrame.

**Exercice 6 (Français)**

* Obtenir toutes les lignes entre la 6ème et la 10ème (incluses) du DataFrame `df_samples`.

### Access rows starting from the 6th row up to and including the 10th row in the df_samples DataFrame.

In [ ]:
# Access rows from the 6th to the 10th row in the df_samples DataFrame
rows_6_to_10 = df_samples.iloc[5:10]

# Print the resulting DataFrame
print(rows_6_to_10)


  sample_id partner_sample_id contributor country location  year  month  \
5  AR0080-C            LUA080  Joao Pinto  Angola   Luanda  2009      4   
6  AR0084-C            LUA084  Joao Pinto  Angola   Luanda  2009      4   
7  AR0097-C            LUA097  Joao Pinto  Angola   Luanda  2009      4   
8  AR0072-C            LUA072  Joao Pinto  Angola   Luanda  2009      4   
9  AR0094-C            LUA094  Joao Pinto  Angola   Luanda  2009      4   

   latitude  longitude sex_call  ... admin1_name admin1_iso  admin2_name  \
5    -8.884     13.302        F  ...      Luanda     AO-LUA       Luanda   
6    -8.884     13.302        F  ...      Luanda     AO-LUA       Luanda   
7    -8.884     13.302        F  ...      Luanda     AO-LUA       Luanda   
8    -8.884     13.302        F  ...      Luanda     AO-LUA       Luanda   
9    -8.884     13.302        M  ...      Luanda     AO-LUA       Luanda   

      taxon  cohort_admin1_year  cohort_admin1_month cohort_admin1_quarter  \
5  coluzzii   

## Indexes

Every pandas DataFrame has an index, which is a special column (or set of columns) which can be used to access rows quickly and conveniently. In particular, indexes can be used to assign a label to every row, and then use that label to locate and access matching rows.

For example, when working with our DataFrame of book characters, we might want to access rows using the name of the character. To do this, we can set the "name" column as the index.

In [ ]:
df_hitchhikers_indexed = df_hitchhikers.set_index("name")
df_hitchhikers_indexed

taxon  gender  number_of_heads  \
name                                                                 
Arthur Dent                         human    male                1   
Tricia McMillan                     human  female                1   
Ford Prefect                 betelgeusian    male                1   
Zaphod Beeblebrox            betelgeusian    male                2   
Marvin the Paranoid Android         robot    None                1   

                             intelligence  
name                                       
Arthur Dent                           100  
Tricia McMillan                       180  
Ford Prefect                          120  
Zaphod Beeblebrox                      30  
Marvin the Paranoid Android       5000000

The DataFrame `df_hitchhikers_indexed` has the same data as before, but notice that the "name" column has moved to the left and is shown in bold. This shows us that the "name" column is now being used as the DataFrame index.

Now that we've set a useful index, we can use the `loc` property to locate rows with a particular name. E.g., locate the row where the character's name is "Tricia McMillan".

In [ ]:
trillian = df_hitchhikers_indexed.loc["Tricia McMillan"]

This indexing action locates a single row, and so pandas returns a Series.

In [ ]:
type(trillian)

pandas.core.series.Series

Inspect the row of data we located.

In [ ]:
trillian

taxon               human
gender             female
number_of_heads         1
intelligence          180
Name: Tricia McMillan, dtype: object

It's also useful to know that you can always reset the index back to the default, using the `reset_index()` function. The default is to index the DataFrame with integers.

In [ ]:
df_hitchhikers_reset = df_hitchhikers_indexed.reset_index()
df_hitchhikers_reset

name         taxon  gender  number_of_heads  \
0                  Arthur Dent         human    male                1   
1              Tricia McMillan         human  female                1   
2                 Ford Prefect  betelgeusian    male                1   
3            Zaphod Beeblebrox  betelgeusian    male                2   
4  Marvin the Paranoid Android         robot    None                1   

   intelligence  
0           100  
1           180  
2           120  
3            30  
4       5000000

**Exercise 7 (English)**

* Set the "sample_id" column as the index on the `df_samples` DataFrame, then locate the row with sample identifier "BK0055-C". Where was this mosquito collected?
* Set the "ID" column as the index on the `df_geneset` DataFrame, then locate the row with identifier "AGAP000818". What is the name of this gene?

**Exercice 7 (Français)**

* Faire de la colonne "sample_id" l'index du DataFrame `df_samples` et ensuite trouver la ligne avec comme identificant d'échantillon "BK0055-C". D'où est issu ce moustique?

* Faire de la colonne "ID" l'index du DataFrame `df_geneset` et ensuite trouver la ligne avec comme identifiant "AGAP000818". Quel est le nom de ce gène?


### Set the "sample_id" column as the index on the df_samples DataFrame, then locate the row with sample identifier "BK0055-C". Where was this mosquito collected?

In [ ]:
# Set "sample_id" as the index in the df_samples DataFrame
df_samples.set_index("sample_id", inplace=True)

# Locate the row with sample identifier "BK0055-C"
row_BK0055_C = df_samples.loc["BK0055-C"]

# Print the location where the mosquito was collected
print("Location of mosquito collection:", row_BK0055_C["country"])


Location of mosquito collection: Central African Republic


### Set the "ID" column as the index on the df_geneset DataFrame, then locate the row with identifier "AGAP000818". What is the name of this gene?

In [ ]:
# Set "ID" as the index in the df_geneset DataFrame
df_geneset.set_index("ID", inplace=True)

# Locate the row with identifier "AGAP000818"
row_AGAP000818 = df_geneset.loc["AGAP000818"]

# Print the name of the gene
print("Name of the gene:", row_AGAP000818["Name"])


Name of the gene: CYP9K1


## Queries

DataFrames also provide a `query()` function, which can be used to select rows matching some condition or combination of conditions. The `query()` function accepts a string containing a query expression.

Let's query the book characters DataFrame to select only rows where the value of the "taxon" column is "human". To match a value we use the "==" operator.


In [ ]:
df_human_hitchhikers = df_hitchhikers.query("taxon == 'human'")
df_human_hitchhikers

name  taxon  gender  number_of_heads  intelligence
0      Arthur Dent  human    male                1           100
1  Tricia McMillan  human  female                1           180

Let's do the opposite, selecting only rows where the value of the "taxon" column is not "human", using the "!=" operator.

In [ ]:
df_non_human_hitchhikers = df_hitchhikers.query("taxon != 'human'")
df_non_human_hitchhikers

name         taxon gender  number_of_heads  \
2                 Ford Prefect  betelgeusian   male                1   
3            Zaphod Beeblebrox  betelgeusian   male                2   
4  Marvin the Paranoid Android         robot   None                1   

   intelligence  
2           120  
3            30  
4       5000000

Let's query a numerical field using the ">" greater than operator.

In [ ]:
df_multi_headed_hitchhikers = df_hitchhikers.query("number_of_heads > 1")
df_multi_headed_hitchhikers

name         taxon gender  number_of_heads  intelligence
3  Zaphod Beeblebrox  betelgeusian   male                2            30

A query expression can combine conditions on multiple columns using "and" and "or". E.g., find rows where the value of the "taxon" column is "human" and the value of the "number_of_heads" column is 1.

In [ ]:
df_non_human_one_headed_hitchhikers = df_hitchhikers.query(
    "taxon != 'human' and number_of_heads == 1"
)
df_non_human_one_headed_hitchhikers

name         taxon gender  number_of_heads  \
2                 Ford Prefect  betelgeusian   male                1   
4  Marvin the Paranoid Android         robot   None                1   

   intelligence  
2           120  
4       5000000

**Exercise 8 (English)**

* Query the `df_samples` DataFrame to select mosquitoes collected in Burkina Faso.
* Query the `df_samples` DataFrame to select mosquitoes collected in Burkina Faso in 2014.
* Query the `df_samples` DataFrame to select Anopheles coluzzii mosquitoes collected in Burkina Faso in 2014 (hint: `taxon == 'coluzzii'`).
* Query the `snp_freqs_df` DataFrame to select SNPs which change the protein sequence (hint: `effect == 'NON_SYNONYMOUS_CODING'`).
* Query the `snp_freqs_df` DataFrame to select SNPs which change the protein sequence and which occur above 5% frequency in at least one cohort (hint: `max_af > 0.05`).

**Exercice 8 (Français)**

* Sélectionner parmi les moustiques du DataFrame `df_samples` ceux qui ont été capturés au Burkina Faso.

* Sélectionner parmi les moustiques du DataFrame `df_samples` ceux qui ont été capturés au Burkina Faso en 2014.

* Sélectionner parmi les moustiques du DataFrame `df_samples` les Anophèles coluzzii qui ont été capturés au Burkina Faso en 2014. (Indice: `taxon == 'coluzzii'`).

* Sélectionner parmi les SNPs du DataFrame `snp_freqs_df` ceux qui modifient la séquence d'une protéine (Indice: `effect == 'NON_SYNONYMOUS_CODING'`).

* Sélectionner parmi les SNPs du DataFrame `snp_freqs_df` ceux qui modifient la séquence d'une protéine et apparaissent avec une fréquence supérieure à 5% dans au moins une cohorte (Indice: `max_af > 0.05`).


### Query the df_samples DataFrame to select mosquitoes collected in Burkina Faso.

In [ ]:
# Query the df_samples DataFrame to select mosquitoes collected in Burkina Faso
mosquitoes_in_burkina_faso = df_samples.query("country == 'Burkina Faso'")

# Print the selected rows
print(mosquitoes_in_burkina_faso)


          partner_sample_id    contributor       country      location  year  \
sample_id                                                                      
AB0085-Cx             BF2-4    Austin Burt  Burkina Faso          Pala  2012   
AB0086-Cx             BF2-6    Austin Burt  Burkina Faso          Pala  2012   
AB0087-C              BF3-3    Austin Burt  Burkina Faso  Bana Village  2012   
AB0088-C              BF3-5    Austin Burt  Burkina Faso  Bana Village  2012   
AB0089-Cx             BF3-8    Austin Burt  Burkina Faso  Bana Village  2012   
...                     ...            ...           ...           ...   ...   
AB0314-C               6775  Nora Besansky  Burkina Faso    Monomtenga  2004   
AB0315-C               6777  Nora Besansky  Burkina Faso    Monomtenga  2004   
AB0316-C               6779  Nora Besansky  Burkina Faso    Monomtenga  2004   
AB0318-C               5072  Nora Besansky  Burkina Faso    Monomtenga  2004   
AB0325-C               1403  Nora Besans

### Query the df_samples DataFrame to select mosquitoes collected in Burkina Faso in 2014.

In [ ]:
import pandas as pd

# Assuming you have already loaded the DataFrame into df_samples

# Query to select mosquitoes collected in Burkina Faso in 2014
selected_samples = df_samples[(df_samples["country"] == "Burkina Faso") & (df_samples["year"] == 2014)]

# Display the selected samples
print(selected_samples)


          partner_sample_id  contributor       country        location  year  \
sample_id                                                                      
AB0326-C             BF18-1  Austin Burt  Burkina Faso    Bana Village  2014   
AB0327-C             BF18-3  Austin Burt  Burkina Faso    Bana Village  2014   
AB0328-C             BF18-4  Austin Burt  Burkina Faso    Bana Village  2014   
AB0329-C             BF18-5  Austin Burt  Burkina Faso    Bana Village  2014   
AB0330-C             BF18-6  Austin Burt  Burkina Faso    Bana Village  2014   
...                     ...          ...           ...             ...   ...   
AB0533-C            BF13-18  Austin Burt  Burkina Faso  Souroukoudinga  2014   
AB0536-C            BF13-31  Austin Burt  Burkina Faso  Souroukoudinga  2014   
AB0537-C            BF13-32  Austin Burt  Burkina Faso  Souroukoudinga  2014   
AB0538-C            BF13-33  Austin Burt  Burkina Faso  Souroukoudinga  2014   
AB0408-C            BF14-20  Austin Burt

### Query the df_samples DataFrame to select Anopheles coluzzii mosquitoes collected in Burkina Faso in 2014 (hint: taxon == 'coluzzii').

In [ ]:
import pandas as pd

# Assuming you have already loaded the DataFrame into df_samples

# Query to select Anopheles coluzzii mosquitoes collected in Burkina Faso in 2014
selected_samples = df_samples[(df_samples["country"] == "Burkina Faso") & (df_samples["year"] == 2014) & (df_samples["taxon"] == "coluzzii")]

# Display the selected samples
print(selected_samples)


          partner_sample_id  contributor       country        location  year  \
sample_id                                                                      
AB0326-C             BF18-1  Austin Burt  Burkina Faso    Bana Village  2014   
AB0327-C             BF18-3  Austin Burt  Burkina Faso    Bana Village  2014   
AB0328-C             BF18-4  Austin Burt  Burkina Faso    Bana Village  2014   
AB0329-C             BF18-5  Austin Burt  Burkina Faso    Bana Village  2014   
AB0330-C             BF18-6  Austin Burt  Burkina Faso    Bana Village  2014   
AB0331-C             BF18-9  Austin Burt  Burkina Faso    Bana Village  2014   
AB0332-C            BF18-10  Austin Burt  Burkina Faso    Bana Village  2014   
AB0334-C            BF18-12  Austin Burt  Burkina Faso    Bana Village  2014   
AB0337-C            BF22-15  Austin Burt  Burkina Faso    Bana Village  2014   
AB0338-C            BF18-13  Austin Burt  Burkina Faso    Bana Village  2014   
AB0339-C            BF18-14  Austin Burt

### Query the snp_freqs_df DataFrame to select SNPs which change the protein sequence (hint: effect == 'NON_SYNONYMOUS_CODING').

In [ ]:
import pandas as pd

# Assuming you have already loaded the DataFrame into snp_freqs_df

# Query to select SNPs that change the protein sequence (effect == 'NON_SYNONYMOUS_CODING')
selected_snps = snp_freqs_df[snp_freqs_df["effect"] == 'NON_SYNONYMOUS_CODING']

# Display the selected SNPs
print(selected_snps)


                                                 pass_gamb_colu_arab  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                      True   
       2362019  G          T          G54C                      True   
       2362023  C          T          P55L                      True   
       2362034  C          A          P59T                      True   
       2362077  G          A          G73D                      True   
...                                                              ...   
       2431328  C          G          A2023G                    True   
       2431371  C          G          S2037R                   False   
       2431417  A          G          I2053V                    True   
       2431424  G          T          G2055V                    True   
       2431613  T          C          V2118A                    True   

                                                 pass_gamb_colu

### Query the snp_freqs_df DataFrame to select SNPs which change the protein sequence and which occur above 5% frequency in at least one cohort (hint: max_af > 0.05).

In [ ]:
import pandas as pd

# The DataFrame already loaded into snp_freqs_df

# Query to select SNPs that change the protein sequence (effect == 'NON_SYNONYMOUS_CODING') and occur above 5% frequency in at least one cohort (max_af > 0.05)
selected_snps = snp_freqs_df[(snp_freqs_df["effect"] == 'NON_SYNONYMOUS_CODING') & (snp_freqs_df["max_af"] > 0.05)]

# Display the selected SNPs
print(selected_snps)


                                                 pass_gamb_colu_arab  \
contig position ref_allele alt_allele aa_change                        
2L     2391228  G          C          V402L                     True   
                           T          V402L                     True   
       2416980  C          T          T791M                     True   
       2422652  A          T          L995F                     True   
       2429617  T          C          I1527T                    True   
       2429745  A          T          N1570Y                    True   
       2429897  A          G          E1597G                    True   
       2429915  A          C          K1603T                    True   
       2430424  G          T          A1746S                   False   
       2430863  T          C          I1868T                    True   
       2430880  C          T          P1874S                    True   
       2430881  C          T          P1874L                    

## Input/output

We saw at the beginning of this module how to read in data from a file. Let's look finally at how to write data out to a file.

To write a DataFrame out to a CSV file, use the [to_csv()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html) function. E.g., let's write out the book characters DataFrame to a CSV file.


In [ ]:
df_hitchhikers.to_csv("hitchhikers.csv")

You can download this file to your local computer by opening the file browser tab to the left in colab.

There are several other file formats you can write to. E.g., you could write to an Excel file. Here, we use the

In [ ]:
df_hitchhikers.to_excel("hitchhikers.xlsx")

Any of the DataFrames we created in this module could be written out in the same way. E.g., we could write a DataFrame that we obtained by querying the book characters for only humans.

In [ ]:
df_human_hitchhikers.to_excel("hh_human.xlsx")

**Exercise 9 (English)**

* Write the `df_samples` DataFrame out to an Excel file.
* Query the `df_samples` DataFrame to select only mosquitoes from Burkina Faso, then write the resulting DataFrame out to a CSV file.

**Exercice 9 (Français)**

* Enregistrer le DataFrame `df_samples` sous forme de fichier Excel.
* Sélectionner parmi les moustiques du DataFrame `df_samples` ceux provenant du Burkina Faso et enregistrer le DataFrame ainsi créé sous forme de fichier CSV.

### Write the df_samples DataFrame out to an Excel file.

In [ ]:
# Write the df_samples DataFrame to an Excel file named "samples_data.xlsx"
df_samples.to_excel("samples_data.xlsx", index=False)


### Query the df_samples DataFrame to select only mosquitoes from Burkina Faso, then write the resulting DataFrame out to a CSV file.

In [ ]:
# Query the df_samples DataFrame to select only mosquitoes from Burkina Faso
mosquitoes_in_burkina_faso = df_samples.query("country == 'Burkina Faso'")

# Write the resulting DataFrame to a CSV file named "mosquitoes_burkina_faso.csv"
mosquitoes_in_burkina_faso.to_csv("mosquitoes_burkina_faso.csv", index=False)


## Topics not covered here

Please note that this module has only covered a few of the many features that pandas supports. Here are a few of the features that we have not covered, but which you might like to learn more about through other online tutorials.

* Grouping and aggregation
* Pivot tables
* Joining and merging
* Concatenating
* Advanced indexing (multi-level indexes; Boolean indexing; ...)
* Modifying data
* Apply
* Missing data
* Time series
* Plotting
* Text
* ...


## Further reading

Here are a few links to online resources providing further information about pandas for new users.

* [pandas getting started](https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html)
* [pandas user guide](https://pandas.pydata.org/docs/user_guide/index.html) (in particular see [10 minutes to pandas](https://pandas.pydata.org/docs/user_guide/10min.html))
* [pandas community tutorials](https://pandas.pydata.org/pandas-docs/stable/getting_started/tutorials.html)